File: skills/inspect-and-debug/harness-doctor/scripts/lib/stats.py
"""Deterministic numeric, grouping, and ranking helpers.

The Python twin of `lib/stats.ts`, function for function, including its
rounding and percentile rules: the two runtimes must produce the same numbers
from the same fact tables, not merely similar ones.
"""

import math
import re
from datetime import datetime, timezone

_WS = re.compile(r"\s+")


def num(value):
    """A finite number, or 0 for None, NaN, infinity, and non-numbers."""
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        return 0
    return value if math.isfinite(value) else 0


def round_to(value, digits=2):
    """Round half up, like JS `Math.round(n * p) / p`, and drop a trailing .0.

    Python's built-in round is half-to-even, so `round(0.5)` is 0 there and 1 in
    JavaScript. Integral results come back as ints so the JSON reads `12`, not
    `12.0`.
    """
    n = num(value)
    p = 10 ** digits
    scaled = math.floor(n * p + 0.5) if n >= 0 else -math.floor(-n * p + 0.5)
    out = scaled / p
    return int(out) if out == int(out) else out


def safe_div(a, b):
    """Divide, returning 0 when the denominator is 0 or not finite."""
    b = num(b)
    return 0 if b == 0 else num(a) / b


def percentile(values, p):
    """Nearest-rank percentile of an unsorted numeric list. 0 when empty."""
    if not values:
        return 0
    ordered = sorted(values)
    idx = min(len(ordered) - 1, max(0, math.ceil(p * len(ordered)) - 1))
    return ordered[idx]


def median(values):
    """Median of an unsorted numeric list. 0 when empty."""
    if not values:
        return 0
    ordered = sorted(values)
    mid = len(ordered) // 2
    if len(ordered) % 2 == 1:
        return ordered[mid]
    return (ordered[mid - 1] + ordered[mid]) / 2


def total(items, f):
    """Sum a numeric field over items. 0 for an empty list."""
    return sum(num(f(x)) for x in items)


def group_by(items, key):
    """Group items into a dict keyed by a string, first-seen order preserved."""
    out = {}
    for item in items:
        out.setdefault(key(item), []).append(item)
    return out


def bump(counts, key, by=1):
    """Increment a counter in a dict."""
    counts[key] = counts.get(key, 0) + by


def uniq_sorted(values):
    """Sorted unique strings."""
    return sorted(set(values))


def top_key(counts):
    """Key with the highest count; ties break on the smallest key."""
    best = ""
    best_n = -1
    for key in sorted(counts):
        n = counts.get(key, 0)
        if n > best_n:
            best = key
            best_n = n
    return best


def sorted_record(entries):
    """Build a dict with keys inserted in sorted order, so JSON is stable."""
    return {k: v for k, v in sorted(entries, key=lambda kv: kv[0])}


def cap_ranking(rows, top):
    """Wrap an already-ordered list in the Ranking shape without re-sorting."""
    cap = max(0, int(top))
    kept = list(rows)[:cap]
    return {"rows": kept, "shown": len(kept), "total": len(rows), "dropped": max(0, len(rows) - len(kept))}


def ranking(rows, metric, tiebreak, top):
    """Sort by metric descending, break ties on tiebreak ascending, then cap."""
    ordered = sorted(rows, key=lambda r: (-num(metric(r)), tiebreak(r)))
    return cap_ranking(ordered, top)


def iso_of(epoch_sec):
    """Epoch seconds as `2026-08-07T14:36:38.000Z`, matching JS toISOString."""
    if epoch_sec is None or not math.isfinite(epoch_sec) or epoch_sec <= 0:
        return ""
    dt = datetime.fromtimestamp(round(epoch_sec), tz=timezone.utc)
    return dt.strftime("%Y-%m-%dT%H:%M:%S.000Z")


def collapse_ws(s):
    """Collapse whitespace runs to single spaces and trim."""
    return _WS.sub(" ", s or "").strip()


def clip(s, max_chars):
    """Cap a string, marking the cut with a single ellipsis char."""
    s = s or ""
    return (s[:max_chars] + "…") if len(s) > max_chars else s


def by_time(rows):
    """Rows with a `t`, sorted ascending. Rows without one are dropped."""
    timed = [r for r in rows if isinstance(r.get("t"), (int, float))]
    return sorted(timed, key=lambda r: r["t"])

File: skills/inspect-and-debug/harness-doctor/scripts/lib/stats.ts
/**
 * Deterministic numeric, grouping, and ranking helpers shared by the aggregate
 * sections and the markdown renderer.
 *
 * Every ranking sorts by the metric, then by a stable string tiebreaker, and every
 * Record is built with its keys inserted in sorted order, so two runs over the same
 * rows produce byte-identical output.
 */

import type { Ranking } from "./types.ts";

/** Sum a numeric field over items. Returns 0 for an empty list. */
export const sum = <T>(items: readonly T[], f: (x: T) => number) =>
  items.reduce((acc, x) => acc + f(x), 0);

/** Round to `digits` decimal places and normalise -0 to 0. */
export const round = (n: number, digits = 2) => {
  if (!Number.isFinite(n)) return 0;
  const p = 10 ** digits;
  const v = Math.round(n * p) / p;
  return Object.is(v, -0) ? 0 : v;
};

/** Divide, returning 0 when the denominator is 0 or not finite. */
export const safeDiv = (a: number, b: number) =>
  b === 0 || !Number.isFinite(b) ? 0 : a / b;

/**
 * Nearest-rank percentile of an unsorted numeric list.
 * `p` is a fraction in [0,1]. Returns 0 for an empty list.
 */
export const percentile = (values: readonly number[], p: number) => {
  if (values.length === 0) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const idx = Math.min(
    sorted.length - 1,
    Math.max(0, Math.ceil(p * sorted.length) - 1),
  );
  return sorted[idx] ?? 0;
};

/** Median of an unsorted numeric list. Returns 0 for an empty list. */
export const median = (values: readonly number[]) => {
  if (values.length === 0) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const mid = sorted.length >> 1;
  return sorted.length % 2 === 1
    ? (sorted[mid] ?? 0)
    : ((sorted[mid - 1] ?? 0) + (sorted[mid] ?? 0)) / 2;
};

/**
 * Group items into a Map keyed by a string.
 * Map insertion order is never used for output: callers rank or sort the entries.
 */
export const groupBy = <T>(items: readonly T[], key: (x: T) => string) => {
  const out = new Map<string, T[]>();
  for (const item of items) {
    const k = key(item);
    const bucket = out.get(k);
    if (bucket) bucket.push(item);
    else out.set(k, [item]);
  }
  return out;
};

/** Increment a counter in a Map. */
export const bump = (counts: Map<string, number>, key: string, by = 1) => {
  counts.set(key, (counts.get(key) ?? 0) + by);
};

/** Plain ascending string compare, locale-independent. */
export const cmpStr = (a: string, b: string) => (a < b ? -1 : a > b ? 1 : 0);

/** Sorted unique strings. */
export const uniqSorted = (values: Iterable<string>) =>
  [...new Set(values)].sort(cmpStr);

/**
 * Key with the highest count. Ties break on the lexicographically smallest key,
 * so the answer never depends on Map insertion order.
 */
export const topKey = (counts: Map<string, number>) => {
  let best = "";
  let bestN = -1;
  for (const key of uniqSorted(counts.keys())) {
    const n = counts.get(key) ?? 0;
    if (n > bestN) {
      best = key;
      bestN = n;
    }
  }
  return best;
};

/** Build a Record with keys inserted in sorted order, so JSON output is stable. */
export const sortedRecord = <T>(entries: readonly (readonly [string, T])[]) => {
  const out: Record<string, T> = {};
  for (const [key, value] of [...entries].sort((a, b) => cmpStr(a[0], b[0]))) {
    out[key] = value;
  }
  return out;
};

/**
 * Sort rows by `metric` descending, break ties on `tiebreak` ascending, then keep
 * the first `top` rows. How many rows were dropped is always reported.
 */
export const ranking = <T>(
  rows: readonly T[],
  metric: (x: T) => number,
  tiebreak: (x: T) => string,
  top: number,
): Ranking<T> => {
  const sorted = [...rows].sort((a, b) => {
    const d = metric(b) - metric(a);
    if (d !== 0) return d;
    return cmpStr(tiebreak(a), tiebreak(b));
  });
  return capRanking(sorted, top);
};

/** Wrap an already-ordered list in the Ranking shape without re-sorting it. */
export const capRanking = <T>(rows: readonly T[], top: number): Ranking<T> => {
  const cap = Math.max(0, Math.floor(top));
  const kept = rows.slice(0, cap);
  return {
    rows: kept,
    shown: kept.length,
    total: rows.length,
    dropped: Math.max(0, rows.length - kept.length),
  };
};

/** Epoch seconds to an ISO-8601 string. Unusable input renders as an empty string. */
export const isoOf = (epochSec: number | undefined) =>
  epochSec === undefined || !Number.isFinite(epochSec) || epochSec <= 0
    ? ""
    : new Date(Math.round(epochSec) * 1000).toISOString();

/** Collapse all whitespace runs to single spaces and trim. */
export const collapseWs = (s: string) => s.replace(/\s+/g, " ").trim();

/** Cap a string to `max` characters, marking the cut with a single ellipsis char. */
export const clip = (s: string, max: number) =>
  s.length > max ? `${s.slice(0, max)}…` : s;

/**
 * Sort rows by their epoch-second field ascending, keeping rows without a
 * timestamp out of the result. Ties keep input order, which is transcript order.
 */
export const byTime = <T extends { t?: number }>(rows: readonly T[]) =>
  rows.filter((r) => typeof r.t === "number").sort((a, b) => (a.t ?? 0) - (b.t ?? 0));
